objective-c/README.md

```diff
@@ -0,0 +1,522 @@
+	+ (BOOL)isSharedIntanceInitialized
+	{
+		NSAssert(self == [MYThisActualClass class], @"Subclasses must implement this method!");
+		
+		return !!sharedInstance;
+	}
+
+	+ (void)initializeSharedInstance
+	{
+		NSAssert(self == [MYThisActualClass class], @"Subclasses must implement this method!");
+		
+		if ([self isSharedIntanceInitialized]) {
+			// This point, the shared instance is already initialized and the assert will fire in DEBUG mode
+			// But just to make sure, skip the re-initializing if the asserts are turned off.
+
+			NSAssert(NO, @"Already initialized!");
+
+			return;
+		}
+
+		sharedInstance = [[self alloc] init];
+	}
+
+	+ (instancetype)sharedInstance
+	{
+		NSAssert(self == [MYThisActualClass class], @"Subclasses must implement this method!");
+		
+		if (![self isSharedIntanceInitialized]) {
+			// This point, the shared instance is hasn't been initialized and the assert will fire in DEBUG mode
+			// But let's initialize the shared instance if the asserts are turned off.
+
+			NSAssert(NO, @"Shared instance is not initialized but trying to access");
+
+			[self initializeSharedInstance];
+		}
+
+		return sharedInstance;
+	}
+
+	- (id)init
+	{
+		self = [super init];
+		if (self) {
+			[[NSNotificationCenter defaultCenter] addObserver:self
+			selector:@selector(handleOtherClassDidSomethingNotification:)
+			name:MYOtherClassDidSomethingNotification
+			object:nil];
+		}
+
+		return self;
+	}
+
+	// Not necessary in a singleton, but recommended
+	- (void)dealloc
+	{
+		[[NSNotificationCenter defaultCenter] removeObserver:self];
+	}
+	...
+	@end
+
+## Notification handlers
+
+Subscribe notifications for your instance but don't forget to unsubscribe. 
+The last place you can do that is the _-dealloc_.
+
+	- (id)init
+	{
+		self = [super init];
+		if (self) {
+			[[NSNotificationCenter defaultCenter] addObserver:self
+			selector:@selector(handleOtherClassDidSomethingNotification:)
+			name:MYOtherClassDidSomethingNotification
+			object:nil];
+		}
+
+		return self;
+	}
+
+	- (void)dealloc
+	{
+		[[NSNotificationCenter defaultCenter] removeObserver:self];
+	}
+
+> Use dedicated methods for handling a notification! Avoid to bind methods like _-clearTableView_.
+> The recommended naming is _-handle[notificatio_name_without_prefix]:_ and always add the 
+> parameter - even if it's not in use.
+
+## Key-value observers' handlers
+
+Use the context parameter when subscribing for a KVO.
+
+Example:
+
+MYClass.m
+
+	@interface MYClass ()
+	{
+		id otherObject;
+	}
+
+	@end
+
+	static void *kvoContext = &kvoContext;
+
+	@implementation MYClass
+
+	#pragma mark - Initialization
+
+	- (id)init
+	{
+		[self = super init];
+		if (self) {
+			otherObject = ...
+			[otherObject addObserver:self 
+			forKeyPath:@"parameter"
+			options:(NSKeyValueObservingOptions)0
+			context:kvoContext];
+	}
+
+		return self;
+	}
+
+	- (void)dealloc
+	{
+		[otherObject removeObserver:self forKeyPath:@"parameter" context:kvoContext];
```

Comment:
`dealloc` is not always a safe place to unregister. In some cases NSNotificationCenter will retain a reference to the object you register and so `dealloc` will never run until you unregister. I think this is mostly fixed in iOS 9, but like I said above, keep in mind on iOS 8 and lower.